include/swift/Reflection/ReflectionContext.h

```diff
@@ -190,8 +193,112 @@ class ReflectionContext
 
     return true;
   }
-#endif // defined(__APPLE__) && defined(__MACH__)
-  
+#else // ELF platforms.
+  bool addImage(RemoteAddress ImageStart) {
+    auto Buf = this->getReader().readBytes(ImageStart, sizeof(Elf64_Ehdr));
+
+    // Read the header.
+    auto Hdr = reinterpret_cast<const Elf64_Ehdr *>(Buf.get());
+
+    if (!Hdr->checkMagic())
+      return false;
+
+    // From the header, grab informations about the section header table.
+    auto SectionHdrAddress = ImageStart.getAddressData() + Hdr->e_shoff;
+    auto SectionHdrNumEntries = Hdr->e_shnum;
+    auto SectionEntrySize = Hdr->e_shentsize;
+
+    // Collect all the section headers, we need them to look up the
+    // reflection sections (by name) and the string table.
+    std::vector<const Elf64_Shdr *> SecHdrVec;
+    for (unsigned I = 0; I < SectionHdrNumEntries; ++I) {
+      auto SecBuf = this->getReader().readBytes(
+          RemoteAddress(SectionHdrAddress + (I * SectionEntrySize)),
+          SectionEntrySize);
+      auto SecHdr = reinterpret_cast<const Elf64_Shdr *>(SecBuf.get());
+      SecHdrVec.push_back(SecHdr);
+    }
+
+    // This provides quick access to the section header string table index.
+    // We also here handle the unlikely even where the section index overflows
+    // and it's just a pointer to secondary storage (SHN_XINDEX).
+    uint32_t SecIdx = Hdr->e_shstrndx;
+    if (SecIdx == SHN_XINDEX) {
+      assert(!SecHdrVec.empty() && "malformed ELF object");
+      SecIdx = SecHdrVec[0]->sh_link;
+    }
+
+    assert(SecIdx < SecHdrVec.size() && "malformed ELF object");
+
+    const Elf64_Shdr *SecHdrStrTab = SecHdrVec[SecIdx];
+    Elf64_Off StrTabOffset = SecHdrStrTab->sh_offset;
+    Elf64_Xword StrTabSize = SecHdrStrTab->sh_size;
```

Comment:
Both of these are wrong.  `offsetof(Elf_SHdr, sh_offset)` is 0x10 vs 0x18.  `offsetof(Elf_SHdr, sh_size)` is 0x14 vs 0x20.  They are 3 vs 8 bytes as well.